### ISSUE TITLE
Explicitly allow or disallow agents with costs in different frameworks

### ISSUE THREAD

**Issue by nicola-bastianello** [team-decent/decent-bench#275]
At this time we don't explicitly allow or disallaw different agents to have costs in different frameworks:

- In some places we raise when the frameworks are mismatched (like when using `Cost._validate_cost_operation(self, other, check_framework: bool = True)`). This would probably disallow mismatched frameworks when e.g. computing the sum of the local costs.
- In other places, we implicitly assume that all costs have the same framework, like when we set `SumCost.framework` as `return self.costs[0].framework`. This doesn't explicitly disallow mismatched frameworks, but silently ignores it

Overall, I think we should decide for a clear behavior and implement it everywhere: either we allow or we disallow mismatched frameworks, and ensure that the docs clarify this.

Thanks to the iop layer and the Array object, I think it's not impossible to allow mismatched frameworks, but there are some places like `SumCost` where it's not clear what we should do.

**Comment by nicola-bastianello** [comment]
> The interoperability layer does not support the use of two different frameworks in one call. For example, iop.stack cannot be called with a list of values from different frameworks. Personally I would disallow the use of different frameworks but still allow for different cost functions on each agent.

ok, I think this is a good compromise if we want to allow working with mixed cost functions. but then we need to decide how to handle `SumCost.framework`, e.g. converting to a single framework? but that probably defeats the purpose of using costs from different frameworks

**Comment by adrianardv** [comment]
Okey, If I have understood correctly you've talked about 3 different things:

### 1) Networks with agents using different frameworks
Case where agent1.cost.framework != agent2.cost.framework
Right now, this is allowed in construction, though network tensors like weights and adjacency are created in the first agent's framework. Then algorithms fail later when they stack/mix states/messages.

Places where compatibility is assumed by reading metadata from the first agent/client instead of validating all agents:
- `networks.py`: in `P2PNetwork` `weights` and `adjacency`; in `FedNetwork` default server initialization.

So, if we are not going to support mixed framework network object, maybe we should raise an error earlier, at construction? If this is the case, then:
- [ ] Add a network level validation step in `networks.py` that checks every agent has the same `cost.shape`, `cost.framework`, and `cost.device`, then the existing "use first agent" helpers can be kept.
- [ ] Keep the current "use first agent" logic, becuase after validation it becomes safe.

If we were to allow it, then:
- [ ] `send`/ `_send_one` would need receiver-side conversion.
- [ ] Fed algorithm aggregation stacks (`iop.stack`) client updates and therefore assumes one backend. ( As Simon said,  iop.stack cannot be called with a list of values from different frameworks.)
- [ ] Mean and consensus metrics also stack all agetn's x


### 2) Agents in a network with different cost functions (same framework/device):
The codebase right now already allows different agents to carry different `Cost` objects and different concrete cost classes. Practical requirement: all agents optimize the same parameter shape, and share framework/device.


### 3) Composition of costs from different frameworks:
If you build a single cost out of multiple sub-costs, are those sub-costs allowed to come from different array frameworks? This is where the code is inconsistent right now.
- `SumCost` "allows" the sum of costs of different frameworks. However, `SumCost` silently reports the framework/device is the first term's framework. Also, in `SumCost.gradient/hessian()`, sub-results are aggregated with iop.stack, which does not support stacking mixed framework arrays in one call.
- In `LinearRegressionCost/LogisticRegressionCost` generic addition same framework/device is not enforced consistently.
- for a` _CompositeRegularizerCost` to be formed, same framework/device is already enforced (raise error).
- `EmpiricalRegularizedCost`, cost and regularizer need to have the same framework/device. It is already enforced (raise error)

If we explicitly disallow 3):
- [ ] Disallow framework/device mismatches inside a composed cost, `SumCost`. --> This can be done changing `_validate_cost_operation()` so framework and device are checked by default. And explicitly validating in `SumCost.__init__` that all terms have the same shape, framework and device. 
- [ ] Keep framework and device as "first term" because that will now be guaranteed.


About the iop-based costs:
- I agree with simon on keeping backend-bound costs specialized.
- Regularizers are already the closest thing to framework-agnostic costs in the codebase, since their methods are iop-based. But they still have explicit framework/device metadata.
- So if a regularizer is created with matching framework/device, composing it with another cost is already supported by the current design, although for backend-bound costs such as PyTorchCost it is still not the most efficient path compared with native framework-specific regularization.


-----

So, my understanding is that the simplest consistent rule is to enforce backend/device homogeneity both at the network level and at the composed-cost level:
- Disallow mixed-framework/device networks (1)
- Disallow mixed-framework/device composed costs (3)
- Keep different cost classes per agent supported, as long as shape/framework/device match (2)

**Comment by Simpag** [comment]
I agree with @adrianardv. We should disallow mixing of frameworks and devices throughout. There is no real reason to allow for this since everything can be done using the same framework and at best allowing it only reduces performance, while it might not even work in some cases. Tensorflow is always compiled during runtime so I am not even sure how well it would work with mixing frameworks or customizing models. Allowing for different cost classes for each agent should be completely fine as long as they use the same framework/device.

**Comment by nicola-bastianello** [comment]
thanks @adrianardv for the clear summary!

I agree with you both: let's disallow mixing in networks and disallow mixing in composite costs

for iop: it would still be nice to have something like `set_seed` for framework+device. this could be called automatically when costs are generated, and future checks refer to these as the correct framework+device. so we don't ask the users to directly specify framework and device unless they need to

**Comment by Simpag** [comment]
> 
> for iop: it would still be nice to have something like `set_seed` for framework+device. this could be called automatically when costs are generated, and future checks refer to these as the correct framework+device. so we don't ask the users to directly specify framework and device unless they need to

That could work but we would need to modify how we create cost functions a little bit. The first code that can run using iop is cost functions initialization. I’m not completely sure how to do this in python yet but we would need to call this “set backend” method before the constructor is called for cost functions because users may need custom construction logic which may require iop functions if its a generic one. This could serve as the global framework check and we wont need to do it in multiple places. Every cost will call it, the first call sets it and if any other call to it sets a different value we raise an error.

**Comment by nicola-bastianello** [comment]
> That could work but we would need to modify how we create cost functions a little bit. The first code that can run using iop is cost functions initialization. I’m not completely sure how to do this in python yet but we would need to call this “set backend” method before the constructor is called for cost functions because users may need custom construction logic which may require iop functions if its a generic one. This could serve as the global framework check and we wont need to do it in multiple places. Every cost will call it, the first call sets it and if any other call to it sets a different value we raise an error.

maybe we could do it with `__init_subclass__` to wrap all subclasses' `__init__` with code that sets the framework/device

**Comment by Simpag** [comment]
> maybe we could do it with `__init_subclass__` to wrap all subclasses' `__init__` with code that sets the framework/device

That could work but there is no need to wrap the `__init__` method, we just perform the logic inside `__init_subclass__`. From my understanding, `__init_subclass__` is called after object creations so `framework/device` should already be set.

**Comment by adrianardv** [comment]
Nice! let's then disallow those two, and later add the iop idea in the interoperability refactor.

> That could work but there is no need to wrap the __init__ method, we just perform the logic inside __init_subclass__. From my understanding, __init_subclass__ is called after object creations so framework/device should already be set.

One small note: `__init_subclass__` by itself is only called when the subclass is defined, not when instances are created, so if we want per-instance backend registration/checks it would need to wrap `__init__` (or we would call the backend registration explicitly inside constructors).
